python/ray/workflow/workflow_state.py

```diff
@@ -0,0 +1,290 @@
+import asyncio
+
+from collections import deque, defaultdict
+import dataclasses
+from dataclasses import field
+import logging
+from typing import List, Dict, Optional, Set, Deque, Callable
+
+import ray
+from ray.workflow.common import (
+    Workflow,
+    StepID,
+    WorkflowRef,
+    WorkflowStepRuntimeOptions,
+)
+from ray.workflow.workflow_context import WorkflowStepContext
+
+logger = logging.getLogger(__name__)
+
+
+@dataclasses.dataclass
+class TaskExecutionMetadata:
+    submit_time: Optional[float] = None
+    finish_time: Optional[float] = None
+    output_size: Optional[int] = None
+
+    @property
+    def duration(self):
+        return self.finish_time - self.submit_time
+
+
+@dataclasses.dataclass
+class WorkflowExecutionState:
```

Comment:
One way to explain this class well is to introduce a workflow and the status of the workflow and use that to explain each fields. For example:

```
H --> *B* -> C --> A
  |--> *D* -> E -|
```
And it's executing D and B. H has been finished. So for the fields below, that's the value. We may just use a simpler one.

No strong opinion. No need to put too much effort in this PR.